### PR TITLE
support to run `header_filter` and `body_filter` phase for global rules.

### LINKED ISSUE #1338: feature: support to run `header_filter` and `body_filter` phase for global rules
original comments: https://github.com/apache/incubator-apisix/issues/1202#issuecomment-603600053

### PR DESCRIPTION
**Pull request by Akayeshmantha**
### Issues resolved

Fix #1338

### CODE REVIEW

File: lua/apisix.lua [hunk 1]
```diff
@@ -435,11 +435,41 @@ end
...
+    end
+
+    if router.global_rules and router.global_rules.values
+            and #router.global_rules.values > 0
+    then
+        local plugins = core.tablepool.fetch("plugins", 32, 0)
+        for _, global_rule in ipairs(router.global_rules.values) do
+            core.table.clear(plugins)
+            plugins = plugin.filter(global_rule, plugins)
+            run_plugin("header_filter")
```
Comment (spacewander) [inline lua/apisix.lua]: Look like we should pass the `plugins` into `run_plugin`?

**Comment by spacewander** [comment]
@Akayeshmantha 
Maybe we can refactor the way to run global rule in log/header filter/body filter phases to use the same function?

**Comment by Akayeshmantha** [comment]
Sure @spacewander created a new issue #1369 since this ticket mentioned a different issue

**Comment by Akayeshmantha** [comment]
@spacewander @membphis

File: lua/apisix.lua [hunk 2]
```diff
@@ -435,12 +435,42 @@ end
...
+    end
+
+    if router.global_rules and router.global_rules.values
+            and #router.global_rules.values > 0
+    then
+        local plugins = core.tablepool.fetch("plugins", 32, 0)
+        for _, global_rule in ipairs(router.global_rules.values) do
+            core.table.clear(plugins)
+            plugins = plugin.filter(global_rule, plugins)
+            run_plugin("header_filter",plugins,api_ctx)
```
Comment (spacewander) [inline lua/apisix.lua]: Need a space after the comma.

Comment (Akayeshmantha) [reply lua/apisix.lua]: oh sorry fixed it

**Comment by Akayeshmantha** [comment]
@membphis  can we merge this if its ok?

File: lua/apisix.lua [hunk 3]
```diff
@@ -435,12 +435,42 @@ end
...
+
+    if router.global_rules and router.global_rules.values
+            and #router.global_rules.values > 0
+    then
+        local plugins = core.tablepool.fetch("plugins", 32, 0)
+        for _, global_rule in ipairs(router.global_rules.values) do
+            core.table.clear(plugins)
+            plugins = plugin.filter(global_rule, plugins)
+            run_plugin("header_filter", plugins, api_ctx)
+        end
```
Comment (membphis) [inline lua/apisix.lua]: missing release `plugins` to tablepool

https://github.com/apache/incubator-apisix/pull/1364/files#diff-1be8f7892798606be4abee3b0bc37860R492

**Comment by Akayeshmantha** [comment]
@membphis done changes

**Comment by Akayeshmantha** [comment]
@membphis @moonming  if things ok can we merge this so I can work on the other issue related to this 🙂

**Comment by membphis** [comment]
@Akayeshmantha you need to rebase you PR after https://github.com/apache/incubator-apisix/commit/c76e7e3623f871987f006848c6c6abcb59550bdb merged

**Comment by Akayeshmantha** [comment]
@membphis  rebased

**Comment by moonming** [comment]
@spacewander please take a look

**Pull request merged**